tests/test_webapp.py

```
@@ -0,0 +1,462 @@
+        conn = http.client.HTTPConnection("127.0.0.1", server.bound_port, timeout=5)
+        conn.request("GET", "/api/snapshot", headers={"Host": "example.test"})
+        resp = conn.getresponse()
+        assert resp.status == 403
+        conn.close()
+    finally:
+        server.shutdown()
+
+
+def test_server_refuses_non_loopback_bind() -> None:
+    with pytest.raises(ValueError, match="loopback-only"):
+        WebappServer(_FakeEngine(), host="0.0.0.0", port=8050)
```

Ruff (0.16.1)

[error] Possible binding to all interfaces

(S104)